modules/ingestor/src/graph/sbom/common/external.rs

```diff
@@ -0,0 +1,115 @@
+use crate::graph::sbom::ReferenceSource;
+use sea_orm::{ActiveValue::Set, ConnectionTrait, DbErr, EntityTrait};
+use sea_query::OnConflict;
+use trustify_common::db::chunk::EntityChunkedIter;
+use trustify_entity::{
+    sbom_external_node::{self, DiscriminatorType, ExternalType},
+    sbom_node,
+};
+use uuid::Uuid;
+
+/// A discriminator for external documents.
+pub struct Discriminator {
+    pub r#type: DiscriminatorType,
+    pub value: String,
+}
+
+impl Discriminator {
+    pub fn new(r#type: DiscriminatorType, value: String) -> Self {
+        Self { r#type, value }
+    }
+
+    pub fn split(value: Option<Self>) -> (Option<DiscriminatorType>, Option<String>) {
+        match value {
+            Some(value) => (Some(value.r#type), Some(value.value)),
+            None => (None, None),
+        }
+    }
+}
+
+// Creator of packages and relationships.
+pub struct ExternalNodeCreator {
+    sbom_id: Uuid,
+    nodes: Vec<sbom_node::ActiveModel>,
+    externals: Vec<sbom_external_node::ActiveModel>,
+}
+
+impl ExternalNodeCreator {
+    pub fn new(sbom_id: Uuid) -> Self {
+        Self {
+            sbom_id,
+
+            nodes: Default::default(),
+            externals: Default::default(),
+        }
+    }
+
+    pub fn add(
+        &mut self,
+        node_id: &str,
+        ext_document_id: &str,
+        ext_node_id: &str,
+        discriminator: Option<Discriminator>,
+    ) {
+        self.nodes.push(sbom_node::ActiveModel {
+            sbom_id: Set(self.sbom_id),
+            node_id: Set(node_id.into()),
+            name: Set(ext_node_id.into()),
+        });
+
+        let (r#type, value) = Discriminator::split(discriminator);
+
+        self.externals.push(sbom_external_node::ActiveModel {
+            sbom_id: Set(self.sbom_id),
+            node_id: Set(node_id.into()),
+            external_doc_ref: Set(ext_document_id.into()),
+            external_node_ref: Set(ext_node_id.into()),
+            external_type: Set(ExternalType::SPDX),
```

Comment:
do u plan to add commits to handle ::CDX ?

Reply:
Good point.